File: usbTerminal/src/main/java/com/ismet/usbterminal/mainscreen/powercommands/PowerCommandsFactory.java
package com.ismet.usbterminal.mainscreen.powercommands;

import android.app.Dialog;
import android.content.Context;
import android.content.Intent;
import android.os.Handler;
import android.os.Message;
import android.view.Window;
import android.widget.TextView;

import com.ismet.usbterminal.MainActivity;
import com.ismet.usbterminalnew.R;
import com.ismet.usbterminal.data.PowerCommand;
import com.ismet.usbterminal.data.PowerState;
import com.proggroup.areasquarecalculator.api.LibraryContentAttachable;

public abstract class PowerCommandsFactory {

	private Dialog mAlertDialog;

	public abstract boolean moveStateToNext();

	public abstract
	@PowerState
	int nextPowerState();

	public abstract PowerCommand currentCommand();

	public abstract
	@PowerState
	int currentPowerState();

	public void sendRequest(final Context context, final Handler mHandler, final CommandsDeliverer
			commandsDeliverer) {
		final boolean handled;
        MainActivity activity = (MainActivity) context;

		@PowerState int mPowerState = currentPowerState();

		switch (mPowerState) {
			case PowerState.OFF_INTERRUPTING:
				Message message = mHandler.obtainMessage();
				message.what = MainActivity.MESSAGE_INTERRUPT_ACTIONS;
				message.sendToTarget();
				handled = true;
				break;
			case PowerState.OFF_WAIT_FOR_COOLING:
				activity.waitForCooling();

				if(context instanceof LibraryContentAttachable) {
					((LibraryContentAttachable) context).dismissProgress();
				}

				mAlertDialog = new Dialog(context);
				mAlertDialog.requestWindowFeature(Window.FEATURE_NO_TITLE);
				mAlertDialog.setContentView(R.layout.layout_cooling);
				mAlertDialog.getWindow().setBackgroundDrawableResource(android.R.color
						.transparent);

				((TextView) mAlertDialog.findViewById(R.id.text)).setText("  Cooling down.  Do not" +
						" " +
						"switch power off.  Please wait . . . ! ! !    \nSystem will turn off " +
						"automaticaly.");

				mAlertDialog.setCancelable(false);

				mAlertDialog.show();

				handled = true;
				break;
			default:
				final PowerCommand currentCommand = currentCommand();
				if (currentCommand != null) {
					if (commandsDeliverer != null) {
						commandsDeliverer.deliverCommand(currentCommand.getCommand());
					}

					if (!currentCommand.hasSelectableResponses()) {
						int powerState = nextPowerState();

						if (powerState != PowerState.OFF && powerState != PowerState.ON) {
							mHandler.postDelayed(new Runnable() {

								@Override
								public void run() {
									PowerCommand currentCommandNew = currentCommand();
									int currentState = currentPowerState();
									if(currentState != PowerState.OFF && currentState !=
											PowerState.ON) {
										if (currentCommand.equals(currentCommandNew)) {
											moveStateToNext();
											currentState = currentPowerState();
											if (currentState != PowerState.OFF && currentState != PowerState.ON) {
												sendRequest(context, mHandler, commandsDeliverer);
											}
										}
									}
								}
							}, currentCommand.getDelay());
						}
					}
					handled = true;
				} else {
					handled = true;
				}
		}

		if (!handled) {
			throw new IllegalArgumentException();
		}
	}

	public Dialog getCoolingDialog() {
		return mAlertDialog;
	}
}

File: usbTerminal/src/main/java/com/ismet/usbterminal/mainscreen/powercommands/PowerCommandsFactory.kt
package com.ismet.usbterminal.mainscreen.powercommands

import android.app.Dialog
import android.os.Handler
import android.view.View
import android.view.Window
import android.widget.TextView
import com.ismet.usbterminal.MainActivity
import com.ismet.usbterminal.data.PowerCommand
import com.ismet.usbterminal.data.PowerState
import com.ismet.usbterminalnew.R

abstract class PowerCommandsFactory {
    var coolingDialog: Dialog? = null
        private set

    abstract fun moveStateToNext(): Boolean
    abstract fun nextPowerState(): PowerState
    abstract fun currentCommand(): PowerCommand?
    abstract fun currentPowerState(): PowerState

    fun sendRequest(activity: MainActivity, mHandler: Handler) {
        var powerState = currentPowerState()
        when (powerState) {
            PowerState.OFF_INTERRUPTING -> {
                val message = mHandler.obtainMessage()
                message.what = MainActivity.MESSAGE_INTERRUPT_ACTIONS
                message.sendToTarget()
            }
            PowerState.OFF_WAIT_FOR_COOLING -> {
                activity.waitForCooling()
                activity.dismissProgress()
                coolingDialog = Dialog(activity).apply {
                    requestWindowFeature(Window.FEATURE_NO_TITLE)
                    setContentView(R.layout.layout_cooling)
                    window!!.setBackgroundDrawableResource(android.R.color.transparent)
                    (findViewById<View>(R.id.text) as TextView).text =
                        """  Cooling down.  Do not switch power off.  Please wait . . . ! ! !    
System will turn off automaticaly."""
                    setCancelable(false)
                    show()
                }
            }
            else -> {
                val currentCommand = currentCommand()
                if (currentCommand != null) {
                    activity.deliverCommand(currentCommand.command)
                    if (!currentCommand.hasSelectableResponses()) {
                        powerState = nextPowerState()
                        if (powerState !== PowerState.OFF && powerState !== PowerState.ON) {
                            mHandler.postDelayed({
                                val currentCommandNew = currentCommand()
                                var currentState = currentPowerState()
                                if (currentState !== PowerState.OFF && currentState !==
                                    PowerState.ON
                                ) {
                                    if (currentCommand == currentCommandNew) {
                                        moveStateToNext()
                                        currentState = currentPowerState()
                                        if (currentState !== PowerState.OFF && currentState !== PowerState.ON) {
                                            sendRequest(activity, mHandler)
                                        }
                                    }
                                }
                            }, currentCommand.delay)
                        }
                    }
                }
            }
        }
    }
}